doc: clarify behaviour of node-api adjust function (#57463)
nodejs/node: doc/api/n-api.md

* `[in] env`: The environment that the API is invoked under.
* `[in] change_in_bytes`: The change in externally allocated memory that is kept
  alive by JavaScript objects.

> * `[out] result`: The adjusted value. Their is no guarrantee that a request to

Reviewer: ```suggestion
* `[out] result`: The adjusted value. There is no guarantee that a request to
```

  which was been adjusted based on calls to `napi_adjust_external_memory` so far.

> For example, if there has been a request for +1000 and a request for -500 it should

Author: ```suggestion
  For example, if there has been a request for +1000 and a request for -500 it should
```

>   For example if there has been a request for +1000 and a request for -500 it should

Author: ```suggestion
For example, if there has been a request for +1000 and a request for -500 it should
```

>   most appropriate.

Reviewer 2: I'm concerned that this renders `napi_adjust_external_memory` as unreliable. It has already been stated that `napi_adjust_external_memory` does not guarantee to trigger GC. And the whole point of this function is to account externally allocated memory held by JS objects.

This `napi_adjust_external_memory` function is also expected to be invoked in a respective "increase a value" and "decrease a same value" at the time of external memory allocation and deallocation. If `napi_adjust_external_memory` increase a value does not result in requested value increase, e.g. requested increase 128KB, but the engine is free to increase 100KB, how is the add-onn authors expected to perform the decrease operation? Should them call `napi_adjust_external_memory ( - 128KB)` later at deallocation, or should them call `napi_adjust_external_memory ( - 100KB)`?

This function should be able to correctly account the value, and it hints the engine to perform GC optimisically. We should not document that it fails its primary motivation.

Reviewer 2: I think for such a runtime, the problem is that how can it handle the symmetry of invocations of `napi_adjust_external_memory`. An add-on is supposed to invoke `napi_adjust_external_memory(-X)` if it calls it `napi_adjust_external_memory(X)` before.

Author: @legendecas I think we can defer that to the runtime. The requirement would be to keep a count that reflects the adjust calls so that the adjusted value returned reflects the + and - requests. The runtime could still decide to totally ingore the requests as it is just a hint. Behind the scenes it could decide not to do anything until a certain total - request.  The main requirement we would be adding is that it maintain a count that reflects the total + and - and return that when the method is called. It would still be a runtime implementation detail in respect to if/how it actually uses the request.

Reviewer 2: I agree with https://github.com/nodejs/node/pull/57463#discussion_r1999094926, and this is also the status-quo. A runtime is not required to do anything (like GC) other than accounting.

The part that concerns me is the following sentence in this PR:

> There is no guarantee that a request to
  adjust by X bytes will result in an adjusted value being X bytes larger than
  before the request.

This breaks the symmetry of `napi_adjust_external_memory(-X)` and `napi_adjust_external_memory(X)`.

Author: @legendecas updated to reflect our discussion.

> * `[out] result`: The adjusted value. This value should reflect the total amount
>   which was been adjusted based on calls to `napi_adjust_external_memory` so far.
>   For example, if there has been a request for +1000 and a request for -500 it should
>   return 500.

Reviewer 2: Note that this example is not "accurate". Right now, there is a single counter for external memory size. That is to say, every add-on, and Node.js cores, are accounted with a single count.

In the upcoming V8 update, V8 is going to deprecate `v8::Isolate::AdjustAmountOfExternalAllocatedMemory`, in favor of a new API [`ExternalMemoryAccounter`](https://source.chromium.org/chromium/chromium/src/+/main:v8/include/v8-external-memory-accounter.h;l=23), which provide a multi-tenant way for accounting.

```suggestion
* `[out] result`: The adjusted value.
```

In order to prepare for the upcoming updates, we should not put this example in. Instead, we should focus on that "if `napi_adjust_external_memory` is called with +500, it is expected to call -500 at a later time".

Author: > That is to say, every add-on, and Node.js cores, are accounted with a single counter.

That is what I was trying to say. I think we can say "if napi_adjust_external_memory is called with +500, it is expected to call -500 at a later time", but that is separate from saying what the result value will be.  In terms of the result we can try to document something that all runtimes can implement, or say that it is up to the imlementation what the result value might be.  In both cases we could say that the expectation is that "if napi_adjust_external_memory is called with +500, it is expected to call -500 at a later time"

I think there may be a few different things to agree on

1) Do calls have to be matched. ie if I make 2 calls of +500 do I need to make 2 calls of -500 to balance out or would one call of -1000 be ok.  My take is that there is no requirement for matching + and - calls but I also can see how that would make it easier for the runtime to do the right thing.

2) How we should document what the value of result should be. Your example of V8 changing its implementation is a good example of how we might get in trouble by being too specific in what we document which is one of the reasons I started off with text that said its is up to the runtime to decide what it returns.  

@legendecas maybe if you suggest an update which reflects what you think we should say for the result that would be a good next step and then we can discuss more in the Node-api team meeting on Friday.

Reviewer 2: Good point that the call times may not be important to be in symmetry. I suggested a change. I hope it makes sense.

> * `[out] result`: The adjusted value. This value should reflect the total amount
>   which was been adjusted based on calls to `napi_adjust_external_memory` so far.
>   For example, if there has been a request for +1000 and a request for -500 it should
>   return 500.

Reviewer 2: ```suggestion
* `[out] result`: The adjusted value. This value should reflect the total amount
  of accounted external memories. Implementations may use a single counter for all
  addons, or a counter for each addons. The absolute value of the returned value
  should not be depended on.
```

Returns `napi_ok` if the API succeeded.

This function gives the runtime an indication of the amount of externally
allocated memory that is kept alive by JavaScript objects
(i.e. a JavaScript object that points to its own memory allocated by a
native addon). Registering externally allocated memory may, but is not

> guarranteed to, trigger global garbage collections more

Reviewer: ```suggestion
guaranteed to, trigger global garbage collections more
```

> often than it would otherwise.

Reviewer 2: ```suggestion
often than it would otherwise.

This function is expected to be invoked symmetrically. If it is invoked with +512KB,
it is expected to invoke -512KB in sum at a later time.
```

* `[in] env`: The environment that the API is invoked under.
* `[in] change_in_bytes`: The change in externally allocated memory that is kept
  alive by JavaScript objects.

> * `[out] result`: The adjusted value. This value should reflect the
>   total amount of external memory with the given `change_in_bytes` included.
>   Implementations may use a single counter for all addons, or a
>   counter for each addon. The absolute value of the returned value should not
>   be depended on.

Author: ```suggestion
* `[out] result`: The adjusted value. This value should reflect the
  total amount of external memory with the given `change_in_bytes` included.
  The absolute value of the returned value should not  be depended on.
  For example, implementations may use a single counter for all addons, or a
  counter for each addon.
```

Returns `napi_ok` if the API succeeded.

This function gives the runtime an indication of the amount of externally
allocated memory that is kept alive by JavaScript objects
(i.e. a JavaScript object that points to its own memory allocated by a
native addon). Registering externally allocated memory may, but is not
guaranteed to, trigger global garbage collections more
often than it would otherwise.

> This function is expected to be invoked symmetrically. If it is invoked with +512KB,

Author: ```suggestion

```

> This function is expected to be invoked symmetrically. If it is invoked with +512KB,

Author: ```suggestion
This function is expected to be called in a manner such that an
addon does not decrease the external memory more than it has
increased the external memory.
```

* `[in] env`: The environment that the API is invoked under.
* `[in] change_in_bytes`: The change in externally allocated memory that is kept
  alive by JavaScript objects.
* `[out] result`: The adjusted value. This value should reflect the
  total amount of external memory with the given `change_in_bytes` included.
  Implementations may use a single counter for all addons, or a
  counter for each addon. The absolute value of the returned value should not
  be depended on.

Returns `napi_ok` if the API succeeded.

This function gives the runtime an indication of the amount of externally
allocated memory that is kept alive by JavaScript objects
(i.e. a JavaScript object that points to its own memory allocated by a
native addon). Registering externally allocated memory may, but is not
guaranteed to, trigger global garbage collections more
often than it would otherwise.

This function is expected to be called in a manner such that an
addon does not decrease the external memory more than it has
increased the external memory.

> it is expected to be invoked -512KB in sum at a later time.

Author: ```suggestion
```
